# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' Clean genetic markers
#' 
#' \code{clean_mrk} filters and/or recodes genetic markers such as single
#' nucleotide polymorphism markers.
#'
#' @param data Either a \code{matrix} or a \code{data.frame} containing marker
#'             genotype data. Markers must be arranged in rows and genotypes
#'             (i.e. lines, hybrids, clones or individuals) in columns. Marker 
#'             names can be provided as row names or in the first column (see 
#'             argument \code{row_names}).
#' @param na_freq Upper limit for the frequency of missing values.
#' @param rm_mono \code{TRUE} to remove all monomorphic markers.
#' @param maf Lower limit for the minor allele frequency.
#' @param recode \code{TRUE} to recode to numeric code.
#' @param rm_dupes \code{TRUE} to remove duplicate markers. Markers sharing
#'                 identical information with missing values for different 
#'                 genotypes are also considered duplicates.
#' @param keep Character vector containing markers to keep.
#' @param impute \code{TRUE} to impute data with mean (for numeric data) or
#'               mode (for categorical data).
#' @param homo Character vector defining homozygous marker genotypes. Default
#'             is c("A", "C", "G", "T") with "A" standing for homozygous for 
#'             nucleotide A.
#' @param hetero Character vector defining heterozygous marker genotypes.
#'               Default is ("R", "Y", "S", "W", "K", "M") with "R" standing 
#'               for a heterozygous genotype with nucleotides A and G according
#'               to IUPAC code.
#' @param homo_num_min Integer value used for recoding of genotypes homozygous
#'                     for minor allele.
#' @param homo_num_maj Integer value used for recoding of genotypes homozygous
#'                     for major allele.
#' @param hetero_num Integer value used for recoding of heterozygous genotypes.
#' @param row_names Must be \code{TRUE} if marker names are provided as row
#'                  names.
#' @param verbose \code{TRUE} to print results at each step.
#'  
#' @return \code{data.frame} with filtered marker genotypes.
#' 
#' @export
clean_mrk <- function(data, na_freq = NULL, rm_mono = FALSE, maf = NULL, recode = FALSE, rm_dupes = FALSE, keep = NULL, impute = FALSE, homo = as.character( c("A", "C", "G", "T")), hetero = as.character( c("R", "Y", "S", "W", "K", "M")), homo_num_min = -1L, homo_num_maj = 1L, hetero_num = 0L, row_names = FALSE, verbose = TRUE) {
    .Call(`_clemag_clean_mrk`, data, na_freq, rm_mono, maf, recode, rm_dupes, keep, impute, homo, hetero, homo_num_min, homo_num_maj, hetero_num, row_names, verbose)
}

